server/src/infra/auth_service.rs

```diff
@@ -314,6 +330,55 @@ where
     get_login_successful_response(&data, &name).await
 }
 
+async fn conventional_login<Backend>(
```

Comment:
Are there no refactoring opportunities for the 2 places that do the entire protocol on the server side?
Maybe returning a result with a custom enum as error for the various cases, cleaning up the code and separating the logic and the error handling.

Reply:
I got no clue how Rust works, but I remember something about error handling in one Rust talk I watched. I will try something, but no guarantee that is any good.